### PR TITLE
ceph: add support for metadata PVC for OSD on PVC

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

We now support the addition of the PVC that acts as a metadata device
for a given OSD.
For this, you need to create a new `volumeClaimTemplates`, its name must
be "metadata" otherwise, Rook won't pick it up.

A template will look like this:

```
volumeClaimTemplates:
- metadata:
    name: data
  spec:
    resources:
      requests:
        storage: 10Gi
    # IMPORTANT: Change the storage class depending on your environment (e.g. local-storage, gp2)
    storageClassName: gp2
    volumeMode: Block
    accessModes:
      - ReadWriteOnce
- metadata:
    name: metadata
  spec:
    resources:
      requests:
        storage: 6Gi
    # IMPORTANT: Change the storage class depending on your environment (e.g. local-storage, gp2)
    storageClassName: gp2
    volumeMode: Block
    accessModes:
      - ReadWriteOnce
```

Closes: https://github.com/rook/rook/issues/3852
Signed-off-by: Sébastien Han <seb@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves https://github.com/rook/rook/issues/3852

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [x] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

Things I'd like to do:

- [x] map both block and block.db as they are in the osd directory (needs c-v modification), we can directly present them
- [x] potentially replace the `activate` initContainer with a simple `ceph-bluetstore-tool` call if we can map the PVC directly in the right location. However, VolumeMounts might conflict...
- [x] Testing operator upgrade
- [x] Testing operator upgrade and then cluster upgrade
- [ ] Testing evicting OSDs

[test ceph]

### CODE REVIEW

File: pkg/operator/ceph/cluster/osd/deviceSet.go [hunk 1]
```diff
@@ -26,72 +26,94 @@ import (
...
-func makeStorageClassDeviceSetPVC(storageClassDeviceSetName, pvcStorageClassDeviceSetPVCId string, pvcIndex, setIndex int, pvcTemplate v1.PersistentVolumeClaim) (pvcs *v1.PersistentVolumeClaim) {
-	pvcLabels := makeStorageClassDeviceSetPVCLabel(storageClassDeviceSetName, pvcStorageClassDeviceSetPVCId, pvcIndex, setIndex)
+func makeStorageClassDeviceSetPVC(storageClassDeviceSetName, pvcStorageClassDeviceSetPVCId string, setIndex int, pvcTemplate v1.PersistentVolumeClaim) (pvcs *v1.PersistentVolumeClaim) {
+	pvcLabels := makeStorageClassDeviceSetPVCLabel(storageClassDeviceSetName, pvcStorageClassDeviceSetPVCId, setIndex)
 
 	// pvc naming format <storageClassDeviceSetName>-<SetNumber>-<PVCIndex>
 	pvcGenerateName := pvcStorageClassDeviceSetPVCId + "-"
 
-	// Add pvcTemplate name to pvc name. i.e <storageClassDeviceSetName>-<SetNumber>-<PVCIndex>-<pvcTemplateName>
-	if len(pvcTemplate.GetName()) != 0 {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/deviceSet.go]: The data and metadata PVC names need to be different. If this is removed it seems the names would be the same. Do you have an example of the PVC names that were generated based on the deviceSet to confirm they are still unique?

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 2]
```diff
@@ -159,9 +162,13 @@ type OrchestrationStatus struct {
...
 type osdProperties struct {
 	//crushHostname refers to the hostname or PVC name when the OSD is provisioned on Nodes or PVC block device, respectively.
-	crushHostname       string
-	devices             []rookalpha.Device
-	pvc                 v1.PersistentVolumeClaimVolumeSource
+	crushHostname string
+	devices       []rookalpha.Device
+	pvc           v1.PersistentVolumeClaimVolumeSource
+	// What about this instead of list for 'pvc' above?
+	// this looks easier, less elegant but not prone to failures as well as upgrades-friendly
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/osd.go]: agreed, a separate property looks good

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 3]
```diff
@@ -252,15 +259,45 @@ func (c *Cluster) startProvisioningOverPVCs(config *provisionConfig) {
...
+		// They are just attached to OSD and field in their property
+		if volume.Type == bluestorePVCMetadata {
+			logger.Infof("PVC %q is not an OSD but a %q device", volume.PersistentVolumeClaimSource.ClaimName, volume.Type)
+			continue
+		}
+
+		// Let's see how the next PVC looks like
+		// If the next PVC has been identified as a PVC let's attached to this OSD property
+		//
+		// The logic will get a bit more complex if we plan support for a third device for "block.wal"
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/osd.go]: Can we not rely on the `VolumeSources` slide ordering to associate the data and metadata PVCs? It feels like we need another struct to help keep them associated correctly. To keep it simple, maybe we add a field to VolumeSources, which is the metadataDevice VolumeSource.

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 4]
```diff
@@ -638,11 +683,23 @@ func (c *Cluster) resolveNode(nodeName string) *rookalpha.Node {
 }
 
 func (c *Cluster) getOSDPropsForPVC(pvcName string) (osdProperties, error) {
-	for _, volumeSource := range c.ValidStorage.VolumeSources {
+	var metadataDevicePVCSource v1.PersistentVolumeClaimVolumeSource
+
+	for i, volumeSource := range c.ValidStorage.VolumeSources {
+		// volumeSource should be consistent and the order alway identical so doing the +1 thing shouldn't be too dangerous
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/osd.go]: typo: always

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 5]
```diff
@@ -102,8 +104,12 @@ if [[ "$CV_MODE" == "lvm" ]]; then
 	# remove the temporary directory
 	rm --recursive --force "$TMP_DIR"
 else
+	ARGS=(--device ${DEVICE} --no-systemd --no-tmpfs)
+	if [ -n "$METADATA_DEVICE" ]; then
+		ARGS+=(--block.db ${METADATA_DEVICE})
+	fi
 	# ceph-volume raw mode only supports bluestore so we don't need to pass a store flag
-	ceph-volume "$CV_MODE" activate --device "$DEVICE" --no-systemd --no-tmpfs
+	ceph-volume "$CV_MODE" activate "${ARGS[@]}"
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: This only applies to LVM-based OSDs, right? Were OSDs not working when on non-PVC metadata devices? Or why is this change needed?

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 6]
```diff
@@ -117,16 +123,25 @@ const (
...
 	if err != nil {
 		return nil, err
 	}
 
-	if osdProps.pvc.ClaimName == "" {
+	if !osdOnPVC {
 		podSpec.Spec.NodeSelector = map[string]string{v1.LabelHostname: osdProps.crushHostname}
 	} else {
+		// This is not needed in raw mode and 14.2.8 brings it
+		// but we still want to do this not to loose backward compatibility with lvm based OSDs...
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: typo: lose

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 7]
```diff
@@ -117,16 +123,25 @@ const (
...
 		podSpec.Spec.InitContainers = append(podSpec.Spec.InitContainers, c.getPVCInitContainer(osdProps.pvc))
+		if osdOnPVCMetadata {
+			// FIX ME
+			// change /srv to something else?
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: why is it using /srv or what should it be?

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 8]
```diff
@@ -567,6 +608,9 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
 // Currently we can't mount a block mode pv directly to a privileged container
 // So we mount it to a non privileged init container and then copy it to a common directory mounted inside init container
 // and the privileged provision container.
+
+// DON'T USE SYMLINK BUT DIRECTLY MOUNT THE BLOCK IN THE CEPH PATH?
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: yes, hoping for this change!

File: Documentation/ceph-cluster-crd.md [hunk 9]
```diff
@@ -81,6 +81,52 @@ spec:
...
+          storageClassName: gp2
+          volumeMode: Block
+          accessModes:
+            - ReadWriteOnce
+      - metadata:
+          name: metadata
+        spec:
+          resources:
+            requests:
+              storage: 5Gi
```
[the remark this answers is not in the document]
Comment (leseb) [reply Documentation/ceph-cluster-crd.md]: Added below recommendation for sizing.

File: Documentation/ceph-cluster-crd.md [hunk 10]
```diff
@@ -81,6 +81,52 @@ spec:
...
+```
+
+> **NOTE**: Note that Rook only supports two naming convention for a given template:
+
+* "data": represents the main OSD block device, where your data are being stored
+* "metadata": represents the metadata device used to store Bluestore database, so it's a backing device, not an OSD.
+It is recommended to use a faster storage class for the metadata device, faster than the one used for the main OSD block.
+Otherwise having a separate metadata device won't bring much performance improvement.
+
+With the present configuration, each OSD will have its main block allocated a 10GB device as well a 5GB device to act as a bluestore database.
```
[the remark this answers is not in the document]
Comment (leseb) [reply Documentation/ceph-cluster-crd.md]: Added link to ceph sizing guide.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 5 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go]: No this applies to raw mode, look at the condition.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 8 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go]: I need to do a more experiment and this requires a change in c-v

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 7 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go]: Well we need to put the block somewhere, `/mnt` is already used and we can't move to subdir since existing cluster have that. So I put it in `/srv`. Ideally, we would have had `/mnt/data/` and `/mnt/metadata/`.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 8 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go]: No need! Here's the commit that does it: 016a494c19bf6028c03232224dd48dde6986c3db, I made it a separate commit in case we decide not to do it.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 5 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go]: removed.

File: cluster/examples/kubernetes/ceph/cluster-on-pvc-metadata.yaml [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+          resources:
+            requests:
+              storage: 10Gi
+          # IMPORTANT: Change the storage class depending on your environment (e.g. local-storage, gp2)
+          storageClassName: gp2
+          volumeMode: Block
+          accessModes:
+            - ReadWriteOnce
+      # dedicated block device to store bluestore database (block.db)
+      - metadata:
```
[the remark this answers is not in the document]
Comment (travisn) [reply cluster/examples/kubernetes/ceph/cluster-on-pvc-metadata.yaml]: thoughts on combining yamls?

File: pkg/daemon/ceph/osd/daemon.go [hunk 12]
```diff
@@ -380,7 +400,19 @@ func getAvailableDevices(context *clusterd.Context, desiredDevices []DesiredDevi
...
 
 		if deviceInfo != nil {
-			available.Entries[device.Name] = deviceInfo
+			// When running on PVC, we typically have a single device only
+			// So it's fine to name the first entry of the map "data" instead of the PVC name
+			// It is particularly useful when a metadata PVC is used because we need to identify it in the map
+			// So the entry must be named "metadata" so it can accessed later
+			if pvcBacked {
+				if device.Type == "data" {
+					available.Entries["data"] = deviceInfo
```
Comment (travisn) [inline pkg/daemon/ceph/osd/daemon.go]: how about defining consts for `"data"` and `"metadata"`?

File: cluster/examples/kubernetes/ceph/cluster-on-pvc-metadata.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (leseb) [reply cluster/examples/kubernetes/ceph/cluster-on-pvc-metadata.yaml]: Yes, I've thought about this, it just seems easier testing wise to have 2 files instead of one. But yes, we can merge.

File: pkg/daemon/ceph/osd/volume.go [hunk 13]
```diff
@@ -115,7 +117,17 @@ func (a *OsdAgent) configureCVDevices(context *clusterd.Context, devices *Device
 			if a.cluster.CephVersion.IsAtLeast(cephVolumeRawModeMinCephVersion) && !lvBackedPV {
 				// For block mode
 				block = fmt.Sprintf("/mnt/%s", a.nodeName)
-				rawOsds, err = getCephVolumeRawOSDs(context, a.cluster.Name, a.cluster.FSID, block, lvBackedPV)
+
+				// This is hard to determine a potential metadata device here
+				// Also, I don't think (leseb) this code we have run in this condition
+				// I tried several things:
+				//    * evict a node, osd moves, the prepare job was never relaunched ever because we check for the osd deployment and skip the prepare
+				//    * restarted the operator, again the prepare job was not re-run
```
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go]: Yes, it's still unclear to me, I need to trace back all the commits because that code looks unreachable to me too. I guess I'm "afraid" to forget a case we solved and introducing a regression.

So I'd like to keep it around a bit so we are sure, wdyt?

File: pkg/daemon/ceph/osd/volume.go [hunk 14]
```diff
@@ -232,21 +237,45 @@ func (a *OsdAgent) initializeBlockPVC(context *clusterd.Context, devices *Device
...
+		if name == "metadata" {
+			logger.Debugf("device %q is a metadata device, skipping this iteration it will be used in the next one", device.Config.Name)
+			// Don't do this device
+			continue
+		}
+
+		// When running on PVC, the prepare job has a single OSD only so 1 disk
+		// However we can present a metadata device so we need to consume it
+		// This will make the devices.Entries larger than usual
+		if len(devices.Entries) > 1 {
```
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go]: Oh yes, at this time, the map entry wasn't named "metadata" yet, good point :)

File: pkg/daemon/ceph/osd/volume.go [hunk 15]
```diff
@@ -232,21 +237,45 @@ func (a *OsdAgent) initializeBlockPVC(context *clusterd.Context, devices *Device
...
+			metadataDev = true
+			metadataArg = append(metadataArg, []string{"--block.db", devices.Entries["metadata"].Config.Name}...)
+			metadataBlockPath = devices.Entries["metadata"].Config.Name
+		}
+
 		if device.LegacyPartitionsFound {
```
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go]: Correct, looks like a leftover!

File: pkg/operator/ceph/cluster/osd/deviceSet.go (hunk 1 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/deviceSet.go]: I'm removing this because it's now including as part of `makeStorageClassDeviceSetPVCIDNew()`.

```
persistentvolumeclaim "set1-data-0-w4bgt" deleted
persistentvolumeclaim "set1-data-1-8c577" deleted
persistentvolumeclaim "set1-data-2-q4c6p" deleted
persistentvolumeclaim "set1-metadata-0-wt5wt" deleted
persistentvolumeclaim "set1-metadata-1-7vz6c" deleted
persistentvolumeclaim "set1-metadata-2-kppl8" deleted
```

File: Documentation/ceph-cluster-crd.md [hunk 16]
```diff
@@ -81,6 +81,64 @@ spec:
...
+          storageClassName: gp2
+          volumeMode: Block
+          accessModes:
+            - ReadWriteOnce
+      - metadata:
+          name: metadata
+        spec:
+          resources:
+            requests:
+              # Find the right size https://docs.ceph.com/docs/mimic/rados/configuration/bluestore-config-ref/#sizing
```
[the remark this answers is not in the document]
Comment (leseb) [reply Documentation/ceph-cluster-crd.md]: Actually I thought I had linked master, good catch!

File: pkg/operator/ceph/cluster/osd/osd.go (hunk 3 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/osd.go]: Won't this be a recursive inclusion then? `VolumeSources` is just a list of `VolumeSource` so I'm not sure this works unless I missed something.

File: pkg/daemon/ceph/osd/volume.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (travisn) [reply pkg/daemon/ceph/osd/volume.go]: Sure let's set `metadataBlock=""`. I think we could simplify the comment though. Not sure we need the details of everything you tested.

File: pkg/operator/ceph/cluster/osd/deviceSet.go [hunk 17]
```diff
@@ -26,72 +26,94 @@ import (
...
 		for i := 0; i < storageClassDeviceSet.Count; i++ {
-			pvc, err := c.createStorageClassDeviceSetPVC(storageClassDeviceSet, i)
-			if err != nil {
-				config.addError("%v", err)
-				config.addError("OSD creation for storageClassDeviceSet %v failed for count %v", storageClassDeviceSet.Name, i)
-				continue
+			// Check if the volume claim template has PVCs
+			if len(storageClassDeviceSet.VolumeClaimTemplates) == 0 {
+				config.addError("no PVC available for storageClassDeviceSet %q", storageClassDeviceSet.Name)
+				return nil
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/deviceSet.go]: how about `continue` instead of `return`?

File: pkg/operator/ceph/cluster/osd/osd.go (hunk 2 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/osd/osd.go]: shall we remove the comment now since the decision is done?

File: pkg/operator/ceph/cluster/osd/osd.go (hunk 3 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/osd/osd.go]: Let's stick with what you have. As long as the slice order is used consistently we should be fine.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 18]
```diff
@@ -158,25 +172,34 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 	replicaCount := int32(1)
 	volumeMounts := opspec.CephVolumeMounts(provisionConfig.DataPathMap, false)
 	configVolumeMounts := opspec.RookVolumeMounts(provisionConfig.DataPathMap, false)
+	osdOnPVC := osdProps.pvc.ClaimName != ""
+	osdOnPVCMetadata := osdProps.metadataPVC.ClaimName != ""
+	// When running on PVC, OSDs don't need a bindmount on dataDirHostPath, only the monitors do
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go]: are the crash logs written to it?

Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go]: yes, but it has its own volumes, just like the logs.

File: pkg/operator/ceph/cluster/osd/deviceSet.go (hunk 17 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/deviceSet.go]: seems ok
